[WIP] Deduplicate Regexp literals (take 2)

casperisfine opened the pull request:
Ruby ticket: https://bugs.ruby-lang.org/issues/16557

Real world application contain many duplicated Regexp literals.

From a rails/console in Redmine:

```
>> ObjectSpace.each_object(Regexp).count
=> 6828
>> ObjectSpace.each_object(Regexp).uniq.count
=> 4162
>> ObjectSpace.each_object(Regexp).to_a.map { |r| ObjectSpace.memsize_of(r) }.sum
=> 4611957 # 4.4 MB total
>> ObjectSpace.each_object(Regexp).to_a.map { |r| ObjectSpace.memsize_of(r) }.sum - ObjectSpace.each_object(Regexp).to_a.uniq.map { |r| ObjectSpace.memsize_of(r) }.sum
=> 1490601 # 1.42 MB could be saved
```

Here's the to 10 duplicated regexps in Redmine:

```
147: /"/
107: /\s+/
103: //
89: /\n/
83: /'/
76: /\s+/m
37: /\d+/
35: /\[/
33: /./
33: /\\./
```

Any empty Rails application will have a similar amount of regexps.

Since https://bugs.ruby-lang.org/issues/16377 made literal regexps frozen, it is possible to deduplicate literal regexps without changing any semantic.

This patch is heavily inspired by the `frozen_strings` table, but applied to literal regexps.

camilo commented:
I build and ran make test locally, works on my machine :tm:

camilo commented:
orig conflct list:

````
# Unmerged paths:                                                                                                                                                                                                           [1936/1936]
#   (use "git add <file>..." to mark resolution)
#
#       both modified:   configure.in
#       both modified:   eval.c
#       both modified:   ext/thread/thread.c
#       both modified:   lib/rubygems.rb
#       both modified:   test/resolv/test_dns.rb
#       both modified:   test/ruby/test_refinement.rb
#       both modified:   test/ruby/test_string.rb
#       both modified:   test/ruby/test_syntax.rb
#       both modified:   test/thread/test_cv.rb
#       both modified:   test/thread/test_queue.rb
#       both modified:   thread_pthread.c
#       both modified:   version.h
#       both modified:   vm_insnhelper.c
````

methodmissing commented:
Just a minor nitpick regarding testing:

```sh
	"  check:           equals make test test-all" \
	"  test:            ruby core tests" \
	"  test-all:        all ruby tests [TESTS=<test files>]" \
```

Not sure if you implied the test target exclusively in the comment, but check covers more of core, which is useful for catching any cruft earlier.

burke commented:
A+, so is this going to be `2.1.6-shopify1` or `2.1.6-shopify2`?

camilo commented:
@methodmissing oh iamma run check, thanks.

@burke @csfrancis is investigating this looks too big of a diff for a .6 we need to make sure nothing will :boom: if I get :+1: after this then yes.

`2.1.6-shopify2`

methodmissing commented:
3rd failure suspicious too, but still pretty close to the expected growth factor of 2.5:

```sh
# Running tests:

[ 2933/13110] TestDRbCore#test_02_unknown = 0.06 s                                                                                                    
  1) Error:
TestDRbCore#test_02_unknown:
DRb::DRbConnError: execution expired
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:741:in `rescue in block in open'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:735:in `block in open'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:734:in `each'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:734:in `open'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:1216:in `initialize'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:1196:in `new'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:1196:in `open'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:1109:in `block in method_missing'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:1128:in `with_friend'
    /Users/lourens/projects/Shopify/ruby/lib/drb/drb.rb:1108:in `method_missing'
    /Users/lourens/projects/Shopify/ruby/lib/drb/extservm.rb:40:in `block in service'
    /Users/lourens/projects/Shopify/ruby/lib/monitor.rb:211:in `mon_synchronize'
    /Users/lourens/projects/Shopify/ruby/lib/drb/extservm.rb:37:in `service'
    /Users/lourens/projects/Shopify/ruby/test/drb/drbtest.rb:30:in `block in ext_service'
    /Users/lourens/projects/Shopify/ruby/test/drb/drbtest.rb:29:in `ext_service'
    /Users/lourens/projects/Shopify/ruby/test/drb/drbtest.rb:69:in `setup_service'
    /Users/lourens/projects/Shopify/ruby/test/drb/test_drb.rb:7:in `setup'

[ 5976/13110] TestMiniTestUnitTestCase#test_capture_subprocess_io = 0.00 s                                                     = 0.00 s = 0.00 s
  2) Skipped:
TestMiniTestUnitTestCase#test_capture_subprocess_io [/Users/lourens/projects/Shopify/ruby/test/minitest/test_minitest_unit.rb:1396]:
Dunno why but the parallel run of this fails

[ 6338/13110] TestObject#test_copied_ivar_memory_leak = 3.44 s                                                         
  3) Failure:
TestObject#test_copied_ivar_memory_leak [/Users/lourens/projects/Shopify/ruby/test/ruby/test_object.rb:820]:
[ruby-core:64700] [Bug #10191].
rss: 4161536 => 10547200..
Expected 2.534448818897638 to be < 2.5.

[10586/13110] TestSocket#test_udp_server = 0.02 s                                                                                                                 
  4) Error:
TestSocket#test_udp_server:
Errno::ECONNREFUSED: Connection refused - recvmsg(2)
    /Users/lourens/projects/Shopify/ruby/test/socket/test_socket.rb:404:in `recvmsg'
    /Users/lourens/projects/Shopify/ruby/test/socket/test_socket.rb:404:in `block (3 levels) in test_udp_server'
    /Users/lourens/projects/Shopify/ruby/.ext/common/socket.rb:73:in `connect_internal'
    /Users/lourens/projects/Shopify/ruby/.ext/common/socket.rb:140:in `connect'
    /Users/lourens/projects/Shopify/ruby/test/socket/test_socket.rb:396:in `block (2 levels) in test_udp_server'
    /Users/lourens/projects/Shopify/ruby/test/socket/test_socket.rb:395:in `each'
    /Users/lourens/projects/Shopify/ruby/test/socket/test_socket.rb:395:in `block in test_udp_server'
    /Users/lourens/projects/Shopify/ruby/.ext/common/socket.rb:642:in `udp_server_sockets'
    /Users/lourens/projects/Shopify/ruby/test/socket/test_socket.rb:325:in `test_udp_server'

Finished tests in 553.707275s, 23.6768 tests/s, 5037.4668 assertions/s.                                               
13110 tests, 2789282 assertions, 1 failures, 2 errors, 29 skips

ruby -v: ruby 2.1.6p336-shopify (development) [x86_64-darwin14.0]
```

csfrancis commented:
Ok, I've done the merge manually on my side: `git merge v2_1_6` into the 2.1 branch, cleaned up the merge conflicts, and I end up with the same diff.

The merge conflicts are sort of confusing. I'm wondering if it's possibly a whitespace issue? In a lot of them, there was no actual conflict -- `HEAD` had no content.

In any case, I think we should test and ship this. We really need to get on 2.2.

camilo commented:
ok iamma see what the failures are about if it looks good nuff build and :sheep:

camilo commented:
yolomerging here.

methodmissing commented:
+ term the build boxes :-)

camilo commented:
done!... I actually need to build it again

byroot committed:
Deduplicate Regexp literals

Ruby ticket: https://bugs.ruby-lang.org/issues/16557

Real world application contain many duplicated Regexp literals.

From a rails/console in Redmine:

```
>> ObjectSpace.each_object(Regexp).count
=> 6828
>> ObjectSpace.each_object(Regexp).uniq.count
=> 4162
>> ObjectSpace.each_object(Regexp).to_a.map { |r| ObjectSpace.memsize_of(r) }.sum
=> 4611957 # 4.4 MB total
>> ObjectSpace.each_object(Regexp).to_a.map { |r| ObjectSpace.memsize_of(r) }.sum - ObjectSpace.each_object(Regexp).to_a.uniq.map { |r| ObjectSpace.memsize_of(r) }.sum
=> 1490601 # 1.42 MB could be saved
```

Here's the to 10 duplicated regexps in Redmine:

```
147: /"/
107: /\s+/
103: //
89: /\n/
83: /'/
76: /\s+/m
37: /\d+/
35: /\[/
33: /./
33: /\\./
```

Any empty Rails application will have a similar amount of regexps.

Since https://bugs.ruby-lang.org/issues/16377 made literal regexps frozen, it is possible to deduplicate literal regexps without changing any semantic.

This patch is heavily inspired by the `frozen_strings` table, but applied to literal regexps.

casperisfine commented:
@XrXr @tenderlove could I get your eyes on this when you have time?

This is a new approach on https://github.com/ruby/ruby/pull/2859, I never managed to finish it because I was trying to do it with "weak" references, but when I saw https://github.com/ruby/ruby/pull/3259 I realized that maybe it wasn't necessary.

This new implementation is mostly based on the assumption that literal regexps are never freed anyway. Am I wrong? 

The idea is to store literals in the following nested hash `VM-> regexp_literals_cache[regexp.options][regexp.src]`. Note that the the second hash is an ident_hash to distinguish equal strings with different encodings, it relies on the regexp src being fstrings.

It almost seem to simple, am I overlooking something?

casperisfine commented:
Profiling build: https://buildkite.com/shopify/shopify-profiling/builds/347
Baseline: https://buildkite.com/shopify/shopify-profiling/builds/335

before:
```
retained memory by class
-----------------------------------
   9.06 MB  Regexp

retained objects by class
-----------------------------------
     11653  Regexp
```

After:
```
retained memory by class
-----------------------------------
   8.16 MB  Regexp

retained objects by class
-----------------------------------
     10231  Regexp
```

So 1422 regexp instance less, for 0.9MB saved. That less than my original calculations, not sure why. Maybe because some of these duplicated Regexp are not literals?

casperisfine commented:
> That less than my original calculations, not sure why. Maybe because some of these duplicated Regexp are not literals?

To answer my own question, there's indeed quite a few duplicated regexps built with `Regexp.new`. For instance [from journey](https://github.com/rails/rails/blob/cf1881237a3b40c6232cb321bed4f03ccc897cf5/actionpack/lib/action_dispatch/journey/path/pattern.rb#L80).

maximecb commented:
This looks inactive so I'm going to close it for now. Feel free to reopen it if still active.
